src/librustc/middle/traits/coherence.rs

```diff
@@ -27,11 +27,12 @@ use syntax::codemap::{DUMMY_SP, Span};
 #[derive(Copy, Clone)]
 struct InferIsLocal(bool);
 
-/// True if there exist types that satisfy both of the two given impls.
-pub fn overlapping_impls(infcx: &InferCtxt,
-                         impl1_def_id: DefId,
-                         impl2_def_id: DefId)
-                         -> bool
+/// If there are types that satisfy both impls, returns a `TraitRef`
+/// with those types substituted.
```

Comment:
So the thing that makes me wary here is that, because of the `probe`, this trait reference may include unresolved type variables that have since been popped. Basically, it's ok to print it out, but it's not ok to do much else with it, since it contains the equivalent of dangling references. I think instead of using `resolve_type_vars_if_possible`, we should consider using `infer::freshen`, which also resolves type variables, but which replaces unresolved type variables with a `FreshTy`. (Possibly not the best names...) One problem though is that freshen also replaces regions with `'static` which, at least if they are printed out, could be pretty confusing. 

I guess another, easier, option would be to move the `probe` call outside of `overlapping_impls` and into `check_if_impls_overlap`. Never mind, let's do that instead. :)